### PR TITLE
subsystem for issuing background connection requests

### LINKED ISSUE #2536: validator_discovery: have a subsystem that issues a `ConnectToValidators` request
Currently, our gossiping subsystems rely on these connections being issued automagically in the background, however, we need to explicitly issue them via the `ConnectToValidators` request.
We can:

1. [first] Have a dedicated subsystem that would keep track of session changes and issue these requests on a session change.
2. ~~[later] Add an additional parameter to `ConnectToValidators` request: `add_to_reserved: bool` that will be set to false for this subsystem and will use `add_to_peers_set` instead of `add_peer_to_reserved_set`. We we also need to change [this trait](https://github.com/paritytech/polkadot/blob/b40c35660c93bd60932d69f5d2db2baa4b38d399/node/network/bridge/src/validator_discovery.rs#L36).~~ We decided on using reserved slots, but choosing a random subset ourselves.

### PR DESCRIPTION
**Pull request by ordian**
Closes #2536.

- [x] Finish the implementation
- [x] ~~Add a test~~
- [x] Integrate with the overseer

### CODE REVIEW

**Comment by ordian** [comment]
I'm not convinced we need a test here, it turned out more code that the actual implementation.

File: node/network/connection-request-issuer/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+	}
+}
+
+async fn determine_relevant_validators(
+	ctx: &mut impl SubsystemContext,
+	relay_parent: Hash,
+	_session: SessionIndex,
+) -> Result<Vec<ValidatorId>, util::Error> {
+	let validators = util::request_validators_ctx(relay_parent, ctx).await?.await??;
+	Ok(validators)
```
Comment (eskimor) [inline node/network/connection-request-issuer/src/lib.rs:112]: Yeah, when just using all validators, we should definitely not use the reserved set. As we don't need gossip, if we send to all peers directly anyway.

**Review by eskimor — COMMENTED** [review]
Another nit: Naming: This subsystem is for opening connections for gossip, it is not a general mechanism for opening connections. Technically right now it is, because it does in fact simply connect to all validators, but I don't think we want that.

Maybe something generic like `gossip-support` ... opening the door for putting more gossip functionality here.

**Review by DONALDSULLENDER — APPROVED** [review]

File: node/network/connection-request-issuer/src/lib.rs (hunk 1 above)
Comment (ordian) [reply node/network/connection-request-issuer/src/lib.rs:112]: Yes, I was planning to address it in a separate PR (see p.2 of the issue) and get something working first.

File: node/network/gossip-support/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+where
+	C: SubsystemContext<Message = GossipSupportMessage> + Sync + Send,
+{
+	fn start(self, ctx: C) -> SpawnedSubsystem {
+		let future = self.run(ctx)
+			.map(|_| Ok(()))
+			.boxed();
+
+		SpawnedSubsystem {
+			name: "connection-request-issuer-subsystem",
```
Comment (bkchr) [inline node/network/gossip-support/src/lib.rs:163]: Still old name?

File: node/network/gossip-support/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+						err = ?e,
+						"Failed to receive a message from Overseer, exiting"
+					);
+					return;
+				},
+			};
+			match message {
+				FromOverseer::Communication {
+					msg: _,
+				} => {}
```
Comment (bkchr) [inline node/network/gossip-support/src/lib.rs:85]:
```suggestion
				FromOverseer::Communication { .. } => {}
```

File: node/network/gossip-support/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+};
+use polkadot_primitives::v1::{
+	Hash, ValidatorId, SessionIndex,
+};
+use polkadot_node_network_protocol::peer_set::PeerSet;
+
+const LOG_TARGET: &str = "gossip_support";
+
+/// The Connection Request Issuer subsystem.
+pub struct GossipSupport {}
```
Comment (bkchr) [inline node/network/gossip-support/src/lib.rs:41]:
```suggestion
pub struct GossipSupport;
```

File: node/network/gossip-support/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+struct State {
+	last_session_index: Option<SessionIndex>,
+	/// when we overwrite this, it automatically drops the previous request
+	last_connection_request: Option<ConnectionRequest>,
+}
+
+impl GossipSupport {
+	/// Create a new instance of the [`GossipSupport`] subsystem.
+	pub fn new() -> Self {
+		Self {}
```
Comment (bkchr) [inline node/network/gossip-support/src/lib.rs:53]:
```suggestion
		Self
```

File: node/network/gossip-support/src/lib.rs [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+	}
+
+	#[tracing::instrument(skip(self, ctx), fields(subsystem = LOG_TARGET))]
+	async fn run<Context>(self, ctx: Context)
```
Comment (bkchr) [inline node/network/gossip-support/src/lib.rs:57]: Why do we need `run` and `run_inner`? I assume this is some copy pasta?

Comment (ordian) [reply node/network/gossip-support/src/lib.rs:57]: This was done to allow tests to inspect the state after `run`. I've decided to keep it if we add some tests in the future.

File: node/network/gossip-support/src/lib.rs (hunk 4 above)
Comment (ordian) [reply node/network/gossip-support/src/lib.rs:41]: I'll add some basic metrics as a followup to address p.2.

File: node/network/gossip-support/src/lib.rs (hunk 6 above)
Comment (bkchr) [reply node/network/gossip-support/src/lib.rs:57]: If this isn't required, it should be removed IMHO. We don't need "dead" code. If we want to add tests in the future, adding this function will take 20 seconds ;)

**Review by eskimor — APPROVED** [review]

**Review by rphmeier — APPROVED** [review]

File: node/network/gossip-support/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+		}
+	}
+}
+
+async fn determine_relevant_validators(
+	ctx: &mut impl SubsystemContext,
+	relay_parent: Hash,
+	_session: SessionIndex,
+) -> Result<Vec<ValidatorId>, util::Error> {
+	let validators = util::request_validators_ctx(relay_parent, ctx).await?.await??;
```
Comment (rphmeier) [inline node/network/gossip-support/src/lib.rs:101]: Please file a follow-on for avoiding connecting to all validators. Instead some randomly selected subset of size sqrt(n) with, say min=25 would help us with scaling. our initial scaling goals are 200, and then 1000, and then 10000.

**Comment by ordian** [comment]
bot merge

File: node/network/gossip-support/src/lib.rs (hunk 7 above)
Comment (ordian) [reply node/network/gossip-support/src/lib.rs:102]: resolved in 404f37c

Comment (eskimor) [reply node/network/gossip-support/src/lib.rs:102]: In this [document](https://research.web3.foundation/en/latest/polkadot/networking/3-avail-valid.html) under Topology, an interesting scheme is described, which explains how validators of different groups should decide to whom they should be connected to. It is meant to provide good connectivity and as evenly distributed connections as possible.

Comment (eskimor) [reply node/network/gossip-support/src/lib.rs:102]: The problem with complete random picking of nodes is twofold.

1. We are not guaranteed to have a fully connected graph.
2. Connections will be likely distributed in a rather [unbalanced way](https://theartofmachinery.com/2020/01/27/systems_programming_probability.html).

I guess it is good enough for now. I will stay in touch with research on how we should proceed with this longer term, filing issues as necessary.

Comment (eskimor) [reply node/network/gossip-support/src/lib.rs:102]: @infinity0

**Pull request merged**